pcsx2-qt/Debugger/Models/BreakpointModel.cpp

```diff
@@ -0,0 +1,314 @@
+/*  PCSX2 - PS2 Emulator for PCs
+ *  Copyright (C) 2002-2023  PCSX2 Dev Team
+ *
+ *  PCSX2 is free software: you can redistribute it and/or modify it under the terms
+ *  of the GNU Lesser General Public License as published by the Free Software Found-
+ *  ation, either version 3 of the License, or (at your option) any later version.
+ *
+ *  PCSX2 is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ *  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+ *  PURPOSE.  See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with PCSX2.
+ *  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "PrecompiledHeader.h"
+
+#include "BreakpointModel.h"
+
+#include "DebugTools/DebugInterface.h"
+#include "DebugTools/Breakpoints.h"
+#include "DebugTools/DisassemblyManager.h"
+
+#include "QtHost.h"
+#include "QtUtils.h"
+#include <QtWidgets/QMessageBox>
+
+#include <algorithm>
+
+BreakpointModel::BreakpointModel(DebugInterface& cpu, QObject* parent)
+	: QAbstractTableModel(parent)
+	, m_cpu(cpu)
+{
+}
+
+int BreakpointModel::rowCount(const QModelIndex&) const
+{
+	return m_breakpoints.size();
+}
+
+int BreakpointModel::columnCount(const QModelIndex&) const
+{
+	return BreakpointColumns::COLUMN_COUNT;
+}
+
+QVariant BreakpointModel::data(const QModelIndex& index, int role) const
+{
+	if (role == Qt::DisplayRole)
+	{
+		auto bp_mc = m_breakpoints.at(index.row());
+
+		if (const auto* bp = std::get_if<BreakPoint>(&bp_mc))
+		{
+			switch (index.column())
+			{
+				case BreakpointColumns::TYPE:
+					return tr("Execute");
+				case BreakpointColumns::OFFSET:
+					return QtUtils::FilledQStringFromValue(bp->addr, 16);
+				case BreakpointColumns::SIZE_LABEL:
+					return m_cpu.GetSymbolMap().GetLabelString(bp->addr).c_str();
+				case BreakpointColumns::OPCODE:
+					// Note: Fix up the disassemblymanager so we can use it here, instead of calling a function through the disassemblyview (yuck)
+					return m_cpu.disasm(bp->addr, true).c_str();
+				case BreakpointColumns::CONDITION:
+					return bp->hasCond ? QString::fromLocal8Bit(bp->cond.expressionString) : tr("No Condition");
+				case BreakpointColumns::HITS:
+					return tr("--");
+				case BreakpointColumns::ENABLED:
+					return bp->enabled ? tr("Enabled") : tr("Disabled");
+			}
+		}
+		else if (const auto* mc = std::get_if<MemCheck>(&bp_mc))
+		{
+			switch (index.column())
+			{
+				case BreakpointColumns::TYPE:
+				{
+					QString type("");
+					type += mc->cond & MEMCHECK_READ ? tr("Read") : "";
+					type += ((mc->cond & MEMCHECK_BOTH) == MEMCHECK_BOTH) ? ", " : " ";
```

Comment:
Some codacy complaints to take care of, below as well.

Reply:
Resolved.